api/src/main/java/org/apache/iceberg/ChangeScan.java

```diff
@@ -0,0 +1,104 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.util.List;
+
+/**
+ * API for configuring a table scan for changes.
+ */
+public interface ChangeScan extends Scan<TableScan> {
+  /**
+   * Instructs this scan to look for changes starting from a particular snapshot (inclusive).
+   * <p>
+   * If the start snapshot is not configured, it is defaulted to the oldest ancestor
+   * of the end snapshot (inclusive).
+   *
+   * @param fromSnapshotId the start snapshot ID (inclusive)
+   * @return this for method chaining
+   */
+  ChangeScan fromSnapshotInclusive(long fromSnapshotId);
+
+  /**
+   * Instructs this scan to look for changes starting from a particular snapshot (exclusive).
+   * <p>
+   * If the start snapshot is not configured, it is defaulted to the oldest ancestor
+   * of the end snapshot (inclusive).
+   *
+   * @param fromSnapshotId the start snapshot ID (exclusive)
+   * @return this for method chaining
+   */
+  ChangeScan fromSnapshotExclusive(long fromSnapshotId);
+
+  /**
+   * Instructs this scan to look for changes up to a particular snapshot (inclusive).
+   * <p>
+   * If the end snapshot is not configured, it is defaulted to the current table snapshot (inclusive).
+   *
+   * @param toSnapshotId the end snapshot ID (inclusive)
+   * @return this for method chaining
+   */
+  ChangeScan toSnapshot(long toSnapshotId);
+
+  /**
+   * Instructs this scan to assign equality delete files to matching data files,
+   * which may be expensive if equality deletes are not scoped to particular partitions.
+   *
+   * @return this for method chaining
+   */
+  ChangeScan resolveEqualityDeletes();
+
+  /**
+   * Plans scan tasks for all file changes.
+   *
+   * @return this for method chaining
+   */
+  FileChangeIndex planFiles();
+
+  /**
+   * An index for looking up file scan tasks for changes in particular snapshots.
+   */
+  interface FileChangeIndex {
+    /**
+     * Returns a list of snapshots that are covered by this index.
+     */
+    List<Snapshot> snapshots();
+
+    /**
+     * Returns scan tasks for data files added in a given snapshot.
+     */
+    Iterable<FileScanTask> addedDataFiles(long snapshotId);
+
+    /**
+     * Returns scan tasks for data files removed in a given snapshot.
+     */
+    Iterable<FileScanTask> removedDataFiles(long snapshotId);
+
+    /**
+     * Returns scan tasks for data files that have at least one matching delete file added in a given snapshot.
+     */
+    Iterable<FileScanTask> dataFilesWithDeletes(long snapshotId);
```

Comment:
I think we need another method that returns the data files with pos deletes file only, since it is an option to handle eq delete separately.

Reply:
If we resolve equality deletes, then we don't have to differentiate between equality or position deletes.
We can simply use `is_deleted` metadata column for both.

Reply:
would the caller handle the read of this case and the `addedDataFiles` differently?

Reply:
Resolving as no longer applies.